src/utils/loose-equal.js

```diff
@@ -16,44 +25,52 @@ function isObject(obj) {
  * Returns boolean true or false
  */
 function looseEqual(a, b) {
-  if (a === b) return true
-  const isObjectA = isObject(a)
-  const isObjectB = isObject(b)
-  if (isObjectA && isObjectB) {
-    try {
-      const isArrayA = isArray(a)
-      const isArrayB = isArray(b)
-      if (isArrayA && isArrayB) {
-        return (
-          a.length === b.length &&
-          a.every((e, i) => {
-            return looseEqual(e, b[i])
-          })
-        )
-      } else if (a instanceof Date && b instanceof Date) {
-        return a.getTime() === b.getTime()
-      } else if (!isArrayA && !isArrayB) {
-        const keysA = keys(a)
-        const keysB = keys(b)
-        return (
-          keysA.length === keysB.length &&
-          keysA.every(key => {
-            return looseEqual(a[key], b[key])
-          })
-        )
-      } else {
-        /* istanbul ignore next */
+  if (a === b) {
+    return true
+  }
+  if (typeof a !== typeof b) {
+    return false
+  }
+  let validTypesCount = [isDate(a), isDate(b)].filter(Boolean).length
+  if (validTypesCount > 0) {
+    return validTypesCount === 2 ? a.getTime() === b.getTime() : false
+  }
+  validTypesCount = [isFile(a), isFile(b)].filter(Boolean).length
+  if (validTypesCount > 0) {
+    return validTypesCount === 2 ? a === b : false
+  }
+  validTypesCount = [isArray(a), isArray(b)].filter(Boolean).length
+  if (validTypesCount > 0) {
+    return validTypesCount === 2
+      ? a.length === b.length && a.every((e, i) => looseEqual(e, b[i]))
+      : false
+  }
+  validTypesCount = [isObject(a), isObject(b)].filter(Boolean).length
+  if (validTypesCount > 0) {
+    if (validTypesCount === 1) {
+      return false
+    }
+    const aKeysCount = keys(a).length
+    const bKeysCount = keys(b).length
+    if (aKeysCount !== bKeysCount) {
+      return false
+    }
+    if (aKeysCount === 0 && bKeysCount === 0) {
+      return String(a) === String(b)
+    }
+    // Using for loop over `Object.keys()` here since some class
+    // keys are not handled correctly otherwise
+    for (const key in a) {
+      if (
+        [a.hasOwnProperty(key), b.hasOwnProperty(key)].filter(Boolean).length === 1 ||
```

Comment:
Since we know `a` will have `key`, can we just check to see if `b` doesn't have the `key`?

Reply:
I think this is the safest and shortest way to handle this.

Reply:
ie:
```js
if ( !b.hasOwnProperty(key) || !looseEqual(a[key], b[key])) {
  return false
}
```

Reply:
The `a.hasOwnProperty(key)` check is also required in the for loop.

Reply:
ok